File: Rust/src/tokens.rs
use std::fmt::{Display, Formatter, Result};

// Used to represent lexigraphical units within a string of text.
pub enum Token<'a> {
    // Represents a word, e.g. apple, car, exit, etc.
    Word(&'a str),

    // Represents a special (non-word) character.
    Special(&'a str)
}

// Implements the Display trait for tokens to make them easier to log.
impl Display for Token<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Token::Word(value) => write!(f, "Word({value})"),
            Token::Special(value) => write!(f, "Special({value})")
        }
    }
}

File: Haskell/app/Tokens.hs
module Tokens
  ( Token (..),
    printTokens,
    printToken,
  )
where

data Token
  = Word String
  | Special Char
  deriving(Show)

printTokens :: [Token] -> IO ()
printTokens = mapM_ printToken

printToken :: Token -> IO ()
printToken (Word text) =
  let formattedString = "Word: " ++ text
   in putStrLn formattedString
printToken (Special c) =
  let formattedString = "Special: " ++ show c
   in putStrLn formattedString